common/url.go

```diff
@@ -393,34 +396,33 @@ func (c URL) Service() string {
 }
 
 // AddParam will add the key-value pair
-// Not thread-safe
-// think twice before using it.
 func (c *URL) AddParam(key string, value string) {
+	c.paramsLock.RLock()
+	defer c.paramsLock.RUnlock()
 	c.params.Add(key, value)
 }
 
 // AddParamAvoidNil will add key-value pair
-// Not thread-safe
-// think twice before using it.
 func (c *URL) AddParamAvoidNil(key string, value string) {
 	if c.params == nil {
 		c.params = url.Values{}
 	}
 
-	c.params.Add(key, value)
+	c.AddParam(key, value)
 }
 
 // SetParam will put the key-value pair into url
-// it's not thread safe.
-// think twice before you want to use this method
 // usually it should only be invoked when you want to initialized an url
 func (c *URL) SetParam(key string, value string) {
+	c.paramsLock.RLock()
+	defer c.paramsLock.RUnlock()
 	c.params.Set(key, value)
 }
 
 // RangeParams will iterate the params
-// it's not thread-safe
 func (c *URL) RangeParams(f func(key, value string) bool) {
```

Comment:
u should exec the @f outside the lock to limit the lock scope.

```Go
        var (
            flag bool
            key string
            value string
        )
        func() {
        c.paramsLock.RLock()
	defer c.paramsLock.RUnlock()
	   for k, v := range c.params {
		if !f(k, v[0]) {
                         key = k
                         value = v[0]
                         flag = true
			break
		}
	  }
        }()
        if flag {
          f(key, value) 
        }
```

Reply:
test this code, u should remove if flag {f(key, value)}